Utilities/Mergers/include/Mergers/MergerCache.h

```diff
@@ -0,0 +1,94 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef ALICEO2_MERGERCACHE_H
+#define ALICEO2_MERGERCACHE_H
+
+/// \file MergerCache.h
+/// \brief Definition of O2 Merger cache, v1.0
+///
+/// \author Piotr Konopka, piotr.jan.konopka@cern.ch
+
+#include <Framework/InputRecord.h>
+
+#include <TObject.h>
+
+#include <memory>
+#include <vector>
+#include <deque>
+
+namespace o2
+{
+namespace mergers
+{
+
+/// \brief Merger cache to store input objects before merging them.
+class MergerCache
+{
+  /// \brief Cache entry storing one mergable object.
+  struct CacheEntry {
+    std::unique_ptr<TObject, void (*)(TObject*)> obj;
+    bool is_merged = false;
+  };
+
+  /// \biref Cache entry queue storing multiple mergable objects received at one input.
+  struct CacheEntryQueue {
+    std::deque<CacheEntry> deque;
```

Comment:
is this meant to aggregate multiple objects before merging them into one?

Reply:
Yes, that would be the case, when we expect differences of objects as inputs.

Reply:
then why is it a deque? Would it not be easier to put them already in a TCollection since that is what the ROOT API dictates? You have to construct it somewhere anyway...
a deque suggests to me that ordering matters AND that you expect to need fast push/pop on both ends - which is not the case since the collection will be flattened anyway...

Reply:
it seems to me there is no need for two different types of CacheEntry - it just complicates life - ROOT API expects a TCollection in both cases.

Reply:
I received requests to implement merging in a moving window, which I plan to do in later updates. In such case I would need some sort of queue.

I am sorry, but I do not understand what you mean in your comment about two different types of CacheEntry.

Reply:
the queue is already provided to you by the underlying data transport mechanism, requeueing only costs memory (and complexity). A moving window type of view is also assured automatically when updating with deltas. I think trying to cover all corner cases in a single code base makes for a very difficult read :)
Anyway - by two types of CacheEntry I mean you have to types: CacheEntry and CacheEntryQueue - if you use only one (the queue by the looks of it) do remove the dead code.

Reply:
maybe a bit of clarification: input data is completely asynchronous and out of order (meaning the messages are multiplexed randomly between producers) so queuing it in another queue does not help with keeping track of time. It can be monitored though if every producer sends some timing information (which is the case) so you can e.g. construct a time span histogram describing your time window belonging to the merged results.

Reply:
I use both of these types - CacheEntryQueue stores CacheEntry'ies inside the queue. Then, there is a separate CacheEntryQueue for each of DPL inputs (so one per producer).

Having the queue, I could place the entries in the right order and easily "move" the window forward, by dropping entries which are older than x seconds compared to the newest entry.
 
The data transport queue is managed by DPL, so from the perspective of a DPL device I don't think I have freedom of modyfing that part of logic.

Reply:
that is correct, you don't have control over the queue ordering - if you need special ordering that should be handled by the framework, otherwise you pay with resource use. Makes little sense to reorder and re-queue messages in every layer.